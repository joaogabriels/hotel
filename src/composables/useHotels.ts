import { ref, computed, watch } from 'vue'
import { useRoute } from 'vue-router'
import { hotelService } from '../services/hotelService'
import type { Hotel } from '../components/models'

export function useHotels() {
  const route = useRoute()
  const hotels = ref<Hotel[]>([])
  const isFetching = ref(false)
  const totalPages = ref(0)
  const currentPage = ref(1)

  const hasNextPage = computed(() => currentPage.value < totalPages.value)

  const fetchHotels = async () => {
    isFetching.value = true

    try {
      const response = await hotelService.getHotels({
        location: route.query.location as string,
        name: route.query.hotelName as string,
        sortBy: route.query.sortBy as 'recommended' | 'best_rated',
        page: Number(route.query.page || 1),
        perPage: 10,
      })

      hotels.value = response.data
      totalPages.value = response.totalPages
      currentPage.value = response.page
    } catch (error) {
      console.error('Error fetching hotels:', error)
      throw error
    } finally {
      isFetching.value = false
    }
  }

  const fetchNextPage = async () => {
    if (!hasNextPage.value || isFetching.value) return

    await fetchHotels()
  }

  watch(
    () => route.query,
    async () => {
      await fetchHotels()
    },
    { immediate: true },
  )

  return {
    hotels,
    isFetching,
    hasNextPage,
    fetchNextPage,
    currentPage,
    totalPages,
  }
}
