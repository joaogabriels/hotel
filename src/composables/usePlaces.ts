import { ref } from 'vue'

import type { Place } from '../components/models'
import { placeService } from '../services/placeService'

export function usePlaces() {
  const places = ref<Place[]>([])
  const isLoading = ref(false)
  const error = ref<Error | null>(null)

  const fetchPlaces = async () => {
    isLoading.value = true
    error.value = null

    try {
      places.value = await placeService.getPlaces()
    } catch (err) {
      error.value = err as Error
      console.error('Error loading places:', err)
    } finally {
      isLoading.value = false
    }
  }

  const filterPlaces = (searchTerm: string) => {
    if (!searchTerm) return places.value

    const needle = searchTerm.toLowerCase()
    return places.value.filter(
      (place) =>
        place.name.toLowerCase().includes(needle) ||
        place.state.name.toLowerCase().includes(needle),
    )
  }

  return {
    places,
    isLoading,
    error,
    fetchPlaces,
    filterPlaces,
  }
}
