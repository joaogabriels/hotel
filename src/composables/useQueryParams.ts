import { useRoute, useRouter } from 'vue-router'

interface QueryParams {
  location?: string
  hotelName?: string
  sortBy?: string
}

export function useQueryParams() {
  const router = useRouter()
  const route = useRoute()

  const updateQueryParams = async (params: Partial<QueryParams>) => {
    const newQuery = {
      ...route.query,
      ...params,
    }

    Object.keys(newQuery).forEach((key) => {
      if (
        newQuery[key as keyof QueryParams] === undefined ||
        newQuery[key as keyof QueryParams] === null ||
        newQuery[key as keyof QueryParams] === ''
      ) {
        delete newQuery[key as keyof QueryParams]
      }
    })

    await router.push({ query: newQuery })
  }

  return {
    updateQueryParams,
  }
}
