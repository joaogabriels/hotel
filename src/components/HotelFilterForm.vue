<template>
  <q-card flat class="form-card">
    <div class="card-header q-pb-md q-mb-lg">
      <h2 class="text-h5 q-ma-none text-grey-9 q-pt-md q-px-md">Reservar hotel</h2>
    </div>

    <q-form @submit="onSubmit">
      <div class="column q-gutter-y-md q-px-md q-pb-md">
        <div class="input-group">
          <label class="input-label">Destino <span class="text-negative">*</span></label>
          <q-select v-model="selectedPlace" dense borderless use-input hide-selected fill-input input-debounce="0"
            :options="filteredPlaces" @filter="handleFilter"
            :option-label="opt => opt ? `${opt.name}, ${opt.state.shortname}` : ''" class="input-field" clearable>
            <template v-slot:no-option>
              <q-item>
                <q-item-section class="text-grey">
                  Nenhum resultado encontrado
                </q-item-section>
              </q-item>
            </template>

            <template v-slot:option="{ itemProps, opt }">
              <q-item v-bind="itemProps">
                <q-item-section>
                  <q-item-label>{{ opt.name }}, {{ opt.state.name }}</q-item-label>
                </q-item-section>
              </q-item>
            </template>
          </q-select>
        </div>

        <div class="input-group">
          <label class="input-label">Nome do Hotel</label>
          <q-input v-model="hotelName" borderless dense class="input-field" />
        </div>

        <div class="flex justify-end q-mt-md">
          <q-btn :label="hasSubmitted ? 'Alterar busca' : 'Buscar'" type="submit" color="primary" rounded no-caps />
        </div>
      </div>
    </q-form>
  </q-card>
</template>

<script setup lang="ts">
import { ref, onMounted } from 'vue';
import type { Place } from './models';
import filterPlaces from '../utils/filterPlaces';
import { useQueryParams } from '../composables/useQueryParams'
import { useRoute } from 'vue-router';

const places = ref<Place[]>([]);
const selectedPlace = ref<Place | null>(null);
const hotelName = ref('');
const filteredPlaces = ref<Place[]>([]);
const hasSubmitted = ref(false);
const { updateQueryParams } = useQueryParams()
const { query } = useRoute()

const loadPlaces = async () => {
  try {
    const response = await fetch('/place.json');
    places.value = await response.json();
    filteredPlaces.value = places.value;
  } catch (error) {
    console.error('Error loading places:', error);
  }
};

const handleFilter = (val: string, update: (callback: () => void) => void) => {
  update(() => {
    filteredPlaces.value = filterPlaces(places.value, val);
  });
};

const onSubmit = async () => {
  if (selectedPlace.value || hotelName.value) {
    hasSubmitted.value = true;
  } else {
    hasSubmitted.value = false;
  }

  await updateQueryParams({
    location: selectedPlace.value ? `${selectedPlace.value.name}, ${selectedPlace.value.state.shortname}` : '',
    hotelName: hotelName.value || ''
  });
};

onMounted(async () => {
  await loadPlaces();

  if (query.location) {
    selectedPlace.value = places.value.find(place => `${place.name}, ${place.state.shortname}` === query.location) || null;

    hasSubmitted.value = true;
  }

  if (query.hotelName) {
    hotelName.value = query.hotelName as string;

    hasSubmitted.value = true;
  }
});
</script>

<style lang="scss" scoped>
.form-card {
  border-radius: 12px;

  .q-btn::before {
    box-shadow: none;
  }
}

.card-header {
  border-bottom: 1px solid rgba(0, 0, 0, 0.12);
}

.input-group {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.input-label {
  font-size: 12px;
  font-weight: 500;
  color: rgba(0, 0, 0, 0.85);
  padding-left: 4px;
}

.input-field {
  :deep(.q-field__control) {
    background: transparent;
    border: 1px solid rgba(0, 0, 0, 0.12);
    border-radius: 4px;
    min-height: 32px;
    padding: 0 12px;
    transition: all 0.3s ease;

    &:hover {
      border-color: rgba(0, 0, 0, 0.24);
    }

    &:focus-within {
      border-color: var(--q-primary);
    }
  }

  :deep(.q-field__marginal),
  :deep(.q-field__control) {
    height: 32px;
  }

  :deep(.q-field__native),
  :deep(.q-field__input) {
    min-height: 32px;
    padding: 0;
    font-size: 13px;
  }

  :deep(.q-field__bottom) {
    min-height: 12px;
    padding: 2px 8px 0;
    font-size: 11px;
  }
}
</style>