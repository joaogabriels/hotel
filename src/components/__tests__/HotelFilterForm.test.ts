import { flushPromises, mount, shallowMount } from '@vue/test-utils'
import type { Mock } from 'vitest'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
import { ref } from 'vue'
import { useRoute } from 'vue-router'

import { mockPlace } from './mocks'

import * as usePlacesModule from '../../composables/usePlaces'
import * as useQueryParamsModule from '../../composables/useQueryParams'
import HotelFilterForm from '../HotelFilterForm.vue'

installQuasarPlugin()

vi.mock('vue-router', () => ({
  useRoute: vi.fn(),
}))

describe('HotelFilterForm', () => {
  afterEach(() => {
    vi.clearAllMocks()
  })

  beforeEach(() => {
    vi.spyOn(useQueryParamsModule, 'useQueryParams').mockReturnValue({
      updateQueryParams: vi.fn(),
    })

    vi.spyOn(usePlacesModule, 'usePlaces').mockReturnValue({
      places: ref([mockPlace]),
      fetchPlaces: vi.fn(),
      filterPlaces: vi.fn().mockReturnValue([mockPlace]),
      isLoading: ref(false),
      error: ref(null),
    })
    ;(useRoute as Mock).mockReturnValue({
      query: {},
    })
  })

  it('matches snapshot', () => {
    const wrapper = shallowMount(HotelFilterForm)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be able to select a place', async () => {
    const wrapper = mount(HotelFilterForm)

    const placeSelect = wrapper.findComponent({ name: 'q-select' })
    const submitButton = wrapper.findAll('.q-btn')[wrapper.findAll('.q-btn').length - 1]

    await placeSelect.vm.$emit('filter', 'Belo', (cb: () => void) => cb())
    await placeSelect.vm.$emit('update:modelValue', mockPlace)

    await submitButton?.trigger('submit')

    await flushPromises()

    expect(useQueryParamsModule.useQueryParams().updateQueryParams).toHaveBeenCalledWith({
      location: 'Belo Horizonte, MG',
      hotelName: '',
    })
  })

  it('should be able to filter name', async () => {
    const wrapper = mount(HotelFilterForm)

    const hotelNameInput = wrapper.findComponent({ name: 'q-input' })
    const submitButton = wrapper.findAll('.q-btn')[wrapper.findAll('.q-btn').length - 1]

    await hotelNameInput.vm.$emit('update:modelValue', 'Hotel Test')
    await submitButton?.trigger('submit')

    await flushPromises()

    expect(useQueryParamsModule.useQueryParams().updateQueryParams).toHaveBeenCalledWith({
      location: '',
      hotelName: 'Hotel Test',
    })
  })

  it('should be able to clear search', async () => {
    const wrapper = mount(HotelFilterForm)

    const submitButton = wrapper.findAll('.q-btn')[wrapper.findAll('.q-btn').length - 1]

    await submitButton?.trigger('submit')

    await flushPromises()

    const clearButton = wrapper.findAll('.q-btn')[wrapper.findAll('.q-btn').length - 2]

    await clearButton?.trigger('click')

    await flushPromises()

    expect(useQueryParamsModule.useQueryParams().updateQueryParams).toHaveBeenCalledWith({
      location: '',
      hotelName: '',
    })
  })

  it('should be able to fill the form when the query params are present', async () => {
    ;(useRoute as Mock).mockReturnValue({
      query: {
        location: 'Belo Horizonte, MG',
        hotelName: 'Hotel Test',
      },
    })

    const wrapper = mount(HotelFilterForm)

    await flushPromises()

    expect(wrapper.findComponent({ name: 'q-input' }).props('modelValue')).toBe('Hotel Test')
    expect(wrapper.findComponent({ name: 'q-select' }).props('modelValue')).toStrictEqual(mockPlace)
  })
})
