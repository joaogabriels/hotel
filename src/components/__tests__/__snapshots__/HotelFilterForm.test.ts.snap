// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelFilterForm > matches snapshot 1`] = `"<q-card-stub data-v-447b4aad=\\"\\" tag=\\"div\\" square=\\"false\\" flat=\\"true\\" bordered=\\"false\\" class=\\"form-card\\"></q-card-stub>"`;
